bpf/lib/conntrack.h

```diff
@@ -466,6 +467,26 @@ static __always_inline void ipv4_ct_tuple_reverse(struct ipv4_ct_tuple *tuple)
 	ct_flip_tuple_dir4(tuple);
 }
 
+static __always_inline int ipv4_ct_extract_l4_ports(struct __ctx_buff *ctx,
+						    int off,
+						    struct ipv4_ct_tuple *tuple)
+{
+#if defined IPV4_FRAGMENTS
+	void *data, *data_end;
+	struct iphdr *ip4;
+
+	// TODO: Can we optimise this out?
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
+
+	if (unlikely(ipv4_is_fragment(ip4)))
+		return ipv4_handle_fragment(ctx, ip4, off,
+					    (struct ipv4_frag_l4ports *)&tuple->dport);
+#endif
+	/* load sport + dport into tuple */
+	return ctx_load_bytes(ctx, off, &tuple->dport, 4);
```

Comment:
Couldn't this use direct packet copy?

Reply:
For this one though, that would require doing the `revalidate_data()` even if `IPV4_FRAGMENTS` is not set?

Reply:
Ah, right. Probably best to leave it that way until we rework all the direct packet loads then.